### DISCUSSION TITLE
Using a language other than typescript / javascript to build vs code extensions..??

### DISCUSSION THREAD

**Discussion by ninetynin** [microsoft/vscode-discussions#64] [Extension Development QnA] [answered]
I read the vscode extension building documentation and its focused towards building vs code extension with either typescript or javascript...
I want to know if there is any other langauge u can use other than these to build these extensions?
I dont want to learn a new langauge rn because im already learning cpp and doing dart meanwhile and I dont want to do too many things at a time but i want to implement a idea i have for vs code extension thats why ...

**Comment by alefragnani** [comment, marked as answer]
Hi @ninetynin ,

The extension itself must be either TypeScript or JavaScript, at least the portion that will work with the VS Code API. On the other hand, you can use any other language to build your _logic_ and you could use any technology to _talk_ to it.

For instance, let's say you want to build a new Language Server for Dart. You can build it using Dart, C++, Rust and any other language, and access it using the Language Server Protocol (https://code.visualstudio.com/api/language-extensions/language-server-extension-guide#why-language-server). 

You can use a similar approach for any other _logic_ you want to build, on any language, and you decide how your extension (written in TS/JS) will _talk_ to your _service_. You can use regular HTTP, RPC, or any other communication technology, either locally or remotely. It's up to you.

Hope this helps

**Reply by ninetynin** [reply]
" and you decide how your extension (written in TS/JS) will talk to your service "
ya but i need to know ts / js for this right? maybe i will just learn it but not go into depth

**Reply by alefragnani** [reply]
Exactly. You have to learn a bit of JS/TS to create an extension for VS Code.

**Reply by ninetynin** [reply]
Nice thanks for saying....
and btw what about visual studio like I remember seeing C# but i didnt see properly and the main idea of this extension isnt to improve a langauge or something like that its just similar to discord real presence

**Reply by alefragnani** [reply]
Visual Studio is a totally different beast 😄 
Not only the language you should use (C#), but also the API.

BTW, VSCode also supports webview, which in theory, you could embed your web app using Dart. I've seen a few extensions using React and Svelte apps. Just an idea

**Reply by ninetynin** [reply]
> Visual Studio is a totally different beast 😄 Not only the language you should use (C#), but also the API.
> 
> BTW, VSCode also supports webview, which in theory, you could embed your web app using Dart. I've seen a few extensions using React and Svelte apps. Just an idea

Oh ill learn c# in future so maybe ill see later on
and btw wdym by vscode supports webview and like embedding ur app? i didnt exactly understand isnt this extension and how is this related to app

**Comment by ssbarnea** [comment]
You can write language servers in any language you want but that may require considerably more effort than in TS, where you already have a lot code already written and well maintained. Keep in mind that distributing binaries will make the building and installation an order of magnitude more complex than using node. At least 3 operating systems multiplied by at least two CPU architectures (amd64 and arm64) and you can easily hate your decision to not use node.

Ansible DevTeam did consider this more than an year ago as we are mostly a Python shop and still, we decided that to take the safe path and have it written in TS. Now we are looking to hire someone with TS and ext development experience to help us move faster.

Also keep in mind the vscode also works in browser! I doubt you will run these other binaries in browser any time soon. Probably something compiled to WASM may work but I am not sure.

**Reply by ninetynin** [reply]
oh thanks i guess i have to learn typescript at somepoint then

**Reply by alefragnani** [reply]
Yes, TypeScript or JavaScript.

**Comment by illusory0x0** [comment]
@ninetynin 

Any programming language that can target JavaScript is supported, but currently few languages ​​have community-driven vscode extension API bindings. I'm currently maintaining the MoonBit vscode extension API binding. You can try using MoonBit to write vscode extensions.

vscode.mbt is in a very early stage, contributions are welcome

repo: [vscode.mbt](https://github.com/moonbit-community/vscode.mbt)

Other programming languages ​​that support targeting JavaScript.

- [OCaml](https://github.com/ocsigen/js_of_ocaml)

- [CoffeeScript](https://coffeescript.org/#introduction)

- [ClojureScript](https://clojurescript.org/)

- [Reason ML](https://reasonml.github.io/en/)

- [PureScript](https://www.purescript.org/)

- [ReScript](https://rescript-lang.org/)

Finally, I wish we had more choices in programming languages.

**Reply by NazmusSayad** [reply]
These are kinda JavaScirpt-ish...
I was thinking if WebAssembly could be used a few days ago. 
No idea how much practical that is for VS Code extension.
For canvas/complex things this can really help for the calculation part.
